File: Medium/1004 - MaxConsecutiveOnesIII/Solution.cs
// Question 1004. Max Consecutive Ones III (https://leetcode-cn.com/problems/max-consecutive-ones-iii/)

/// <summary>
/// 执行用时：216 ms, 在所有 C# 提交中击败了 52.94% 的用户
/// 内存消耗：39 MB, 在所有 C# 提交中击败了 41.18% 的用户
/// 
/// Acknowledgement: This solution borrows idea from the 
/// [official solution](https://leetcode-cn.com/problems/max-consecutive-ones-iii/solution/fen-xiang-hua-dong-chuang-kou-mo-ban-mia-f76z/)
/// </summary>
public class Solution 
{
    public int LongestOnes(int[] A, int K) 
    {
        int N = A.Count();
        
        int left = 0;
        int right = 0;
        int result = 0;

        while (right < N)
        {
            if (A[right] == 0) 
            {
                K--;
            }

            while (K < 0)
            {
                K += A[left] == 0 ? 1 : 0;
                left++;
            }
            
            result = Math.Max(result, right - left + 1);
            right++;
        }

        return result;
    }
}

File: Medium/1004 - MaxConsecutiveOnesIII/solution.ts
// Question 1004. Max Consecutive Ones III (https://leetcode-cn.com/problems/max-consecutive-ones-iii/)

/**
 * 执行用时：104 ms, 在所有 TypeScript 提交中击败了 50.00% 的用户
 * 内存消耗：43.8 MB, 在所有 TypeScript 提交中击败了 90.00% 的用户
 *
 * Acknowledgement: This solution borrows idea from the 
 * [official solution](https://leetcode-cn.com/problems/max-consecutive-ones-iii/solution/fen-xiang-hua-dong-chuang-kou-mo-ban-mia-f76z/)
 * @param {number[]} A
 * @param {number} K
 * @returns {number}
 */
function longestOnes(A: number[], K: number): number {
  let left: number = 0;
  let right: number = 0;
  let result: number = 0;
  const N: number = A.length;

  while (right < N) {
    if (!A[right]) {
      K--;
    }

    while (K < 0) {
      K += A[left] ? 0 : 1;
      left++;
    }

    result = Math.max(result, right - left + 1);
    right++;
  }

  return result;
}
